### PR TITLE
[String] Introduce a locale-aware Slugger in the String component

### PR DESCRIPTION
**Pull request by tgalopin**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

This PR introduces a locale-aware Slugger in the new String component, based on transliterators provided by the PHP intl extension (or iconv if not available). It also wires this Slugger in the FrameworkBundle in order to use it easily as a service and to automatically inject the proper locale into it to choose the appropriate transliteration depending on the Request locale.

See https://github.com/unicode-org/cldr/tree/master/common/transforms for CLDR mappings.

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 1]
```diff
@@ -194,6 +194,14 @@ public function load(array $configs, ContainerBuilder $container)
             }
         }
 
+        // If the slugger is used but the String component is not available, we should throw an error
+        if (!class_exists('Symfony\Component\String\Slugger\SluggerInterface')) {
+            $definition = new Definition('stcClass');
```
Comment (yceruto) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: `stdClass`

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 2]
```diff
@@ -194,6 +195,12 @@ public function load(array $configs, ContainerBuilder $container)
             }
         }
 
+        // If the slugger is used but the String component is not available, we should throw an error
+        if (!class_exists(SluggerInterface::class)) {
+            $container->register('slugger', 'stdClass')
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: stdClass ?

File: src/Symfony/Component/String/Slugger/AsciiSlugger.php [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        return $this->defaultLocale;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function slug(string $string, string $separator = '-', string $locale = null): AbstractUnicodeString
+    {
+        if (!($locale = $locale ?? $this->defaultLocale) || !\function_exists('transliterator_transliterate')) {
+            return $this->createTransliteratedSlug($string, $separator, $this->transliterators[$locale] ?? null);
```
Comment (stof) [inline src/Symfony/Component/String/Slugger/AsciiSlugger.php]: this will break if `$locale` is null, which is possible if `defaultLocale` is null.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml [hunk 4]
```diff
@@ -120,5 +120,11 @@
             <argument type="service" id="request_stack" />
             <tag name="kernel.event_subscriber" />
         </service>
+
+        <service id="slugger" class="Symfony\Component\String\Slugger\AsciiSlugger">
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml]: should be in a file loaded only when the component class exists.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 5]
```diff
@@ -194,6 +195,12 @@ public function load(array $configs, ContainerBuilder $container)
...
+                ->addError('You cannot use the Slugger since the String component is not installed. Try running "composer require symfony/string".');
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Do we actually use this pattern anywhere else ?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 2 above)
Comment (tgalopin) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: What's your question?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml (hunk 4 above)
Comment (tgalopin) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml]: Not necessarily: my aim here is to provide a nicer error message when the String component is required. This means replacing this definition with the stdClass definition when the component is not present, asking to install the component. This also means that this definition can be stored here IMO.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 5 above)
Comment (tgalopin) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: I don't think so, because most of the time we have bundle configuration to help us determine whether the feature is used or not. In our case, typehinting is the only usage information we have, thus this definition.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: a class is required, but any class works, so stdClass or whatever :)

**Review by nicolas-grekas — APPROVED** [review]

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/String/Slugger/AsciiSlugger.php [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function slug(string $string, string $separator = '-', string $locale = null): AbstractUnicodeString
+    {
+        $locale = $locale ?? $this->defaultLocale;
+
+        $transliterator = [];
+        if ('de' === $locale || 0 === strpos($locale, 'de_')) {
```
Comment (javiereguiluz) [inline src/Symfony/Component/String/Slugger/AsciiSlugger.php]: For your consideration: I read this code and I don't understand why we need to introduce this exception for the `de` locale. Maybe add a very brief comment explaining why?

File: src/Symfony/Component/String/Slugger/SluggerInterface.php [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ * Creates a URL-friendly slug from a given string.
+ *
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ *
+ * @experimental in 5.0
+ */
+interface SluggerInterface
+{
+    /**
+     * Creates a slug based on a given string and an given locale, using appropriate
```
Comment (javiereguiluz) [inline src/Symfony/Component/String/Slugger/SluggerInterface.php]: `an given locale` -> `a given locale`  ... although I'd reword it a bit to avoid repeating "a given":

`Creates a slug for the given string and locale, using appropriate transliteration when needed.`

**Review by javiereguiluz — APPROVED** [review]
I love this! This will be immensely useful 😍  Thanks Titouan!

File: src/Symfony/Component/String/Slugger/AsciiSlugger.php (hunk 6 above)
Comment (tgalopin) [reply src/Symfony/Component/String/Slugger/AsciiSlugger.php]: I did have a comment then removed it, I'll add it back :) . Basically, transliterating German is simple enough that we can provide a polyfill (which is integrated into the ascii() method) and thus we do not need the transliterator.

**Comment by nicolas-grekas** [comment]
Thank you @tgalopin.
